File: testSuite/regressions/mergerTreeBoxSizeWeight.pl
#!/usr/bin/env perl
use strict;
use warnings;
use PDL;
use PDL::IO::HDF5;

# Run a test case which checks that volume weights of merger trees read from file are assigned correctly when they are computed
# from the simulation box size.
# Andrew Benson (16-January-2014)

# Run the model and check for successful completion.
system("./Galacticus.exe testSuite/regressions/mergerAtFinalTimeInTree.xml");
die("FAILED: mergerTreeBoxSizeWeight.pl model failed to complete") 
    unless ( $? == 0 );

# Extract required information from files.
my $trees                       = new PDL::IO::HDF5("testSuite/data/mergerTrees/mergerAtFinalTimeInTree.hdf5"   );
my $model                       = new PDL::IO::HDF5("testSuite/outputs/regressions/mergerAtFinalTimeInTree.hdf5");
(
 my $boxSize
)                               = $trees->group("simulation")
    ->attrGet(
    'boxSize'
    );
(
 my $lengthHubbleExponent     ,
 my $lengthScaleFactorExponent,
 my $lengthUnitsInSI
)                               = $trees->group("units"     )
    ->attrGet(
    'lengthHubbleExponent'     ,
    'lengthScaleFactorExponent',
    'lengthUnitsInSI'
    );
(
 my $hubble
)                               = $trees->group("cosmology" )
    ->attrGet(
    'HubbleParam'
    );
my $treeWeight                  = $model->group("Outputs"   )->group("Output1")->dataset("mergerTreeWeight")->get();

# Compute expected tree weight.
my $scaleFactor    = pdl 1.0;
my $megaParsec     = pdl 3.08567758135e+22;
$boxSize           = $boxSize*($hubble**$lengthHubbleExponent)*($scaleFactor**$lengthScaleFactorExponent)*($lengthUnitsInSI/$megaParsec);
my $weightExpected = 1.0/$boxSize**3;

# Compute fractional difference.
my $fractionalError = abs($treeWeight-$weightExpected)/$weightExpected;

# Check for consistency.
die("FAILED: mergerTreeBoxSizeWeight.pl tree weights do not equal expected values")
    unless ( all($fractionalError < 1.0e-6) );

exit;

File: testSuite/regressions/mergerTreeBoxSizeWeight.py
#!/usr/bin/env python3
import subprocess
import sys
import h5py
import numpy as np

# Run a test case which checks that volume weights of merger trees read from file are assigned
# correctly when they are computed from the simulation box size.
# Andrew Benson (16-January-2014; ported to Python)

# Run the model and check for successful completion.
status = subprocess.run("./Galacticus.exe testSuite/regressions/mergerAtFinalTimeInTree.xml", shell=True)
if status.returncode != 0:
    print("FAILED: mergerTreeBoxSizeWeight model failed to complete")
    sys.exit(1)

# Extract required information from files.
with h5py.File("testSuite/data/mergerTrees/mergerAtFinalTimeInTree.hdf5", "r") as trees:
    box_size                   = trees["simulation" ].attrs["boxSize"               ]
    length_hubble_exponent      = trees["units"      ].attrs["lengthHubbleExponent"  ]
    length_scale_factor_exponent = trees["units"      ].attrs["lengthScaleFactorExponent"]
    length_units_in_SI          = trees["units"      ].attrs["lengthUnitsInSI"       ]
    hubble                     = trees["cosmology"   ].attrs["HubbleParam"           ]

with h5py.File("testSuite/outputs/regressions/mergerAtFinalTimeInTree.hdf5", "r") as model:
    tree_weight = model["Outputs/Output1/mergerTreeWeight"][:]

# Compute expected tree weight.
scale_factor = 1.0
mega_parsec  = 3.08567758135e+22
box_size     = (box_size
                * hubble**length_hubble_exponent
                * scale_factor**length_scale_factor_exponent
                * length_units_in_SI / mega_parsec)
weight_expected  = 1.0 / box_size**3

# Compute fractional difference.
fractional_error = np.abs(tree_weight - weight_expected) / weight_expected

# Check for consistency.
if np.any(fractional_error >= 1.0e-6):
    print("FAILED: mergerTreeBoxSizeWeight tree weights do not equal expected values")
    sys.exit(1)

print("SUCCESS: mergerTreeBoxSizeWeight")
